File: factory/src/main/java/com/zuzex/factory/service/GrpcOrderCarService.java
package com.zuzex.factory.service;

import com.zuzex.common.dto.OrderDto;
import com.zuzex.common.dto.OrderResultDto;
import com.zuzex.common.grpc.dto.CommonDto;
import com.zuzex.common.grpc.service.CommonServiceGrpc;
import com.zuzex.common.model.EventStatus;
import com.zuzex.factory.mapper.OrderMapper;
import com.zuzex.factory.model.Order;
import io.grpc.stub.StreamObserver;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.lognet.springboot.grpc.GRpcService;

@Slf4j
@GRpcService
@RequiredArgsConstructor
public class GrpcOrderCarService extends CommonServiceGrpc.CommonServiceImplBase {

    private final OrderService orderService;
    private final OrderMapper orderMapper;

    @Override
    public void createNewCarOrder(CommonDto.OrderDto request, StreamObserver<CommonDto.OrderResultDto> responseObserver) {
        OrderDto orderDto = orderMapper.grpcOrderDtoToOrderDto(request);
        log.info("Message '{}' received", orderDto);

        try {
            Order orderAfterSave
                    = orderService.createNewOrderByCarIdAndDescription(orderDto.getCarId(), orderDto.getOrderDescription());
            log.info("Order '{}' saved", orderAfterSave);

            responseObserver.onNext(orderMapper.orderResultDtoToGrpcOrderResultDto(
                    OrderResultDto.builder()
                            .carId(orderAfterSave.getCarId())
                            .eventStatus(EventStatus.SUCCESS)
                            .build()));
        } catch (Exception ex) {
            log.info("Something went wrong... {}", ex.getMessage());

            responseObserver.onNext(orderMapper.orderResultDtoToGrpcOrderResultDto(
                    OrderResultDto.builder()
                            .carId(orderDto.getCarId())
                            .eventStatus(EventStatus.FAILED)
                            .build()));
        } finally {
            responseObserver.onCompleted();
        }
    }
}

File: car-factory/src/main/kotlin/com/zuzex/carfactory/service/GrpcOrderCarService.kt
package com.zuzex.carfactory.service

import com.zuzex.carfactory.mapper.OrderMapper
import com.zuzex.common.grpc.dto.CommonDto
import com.zuzex.common.grpc.service.CommonServiceGrpc
import com.zuzex.common.model.EventStatus
import io.grpc.stub.StreamObserver
import mu.KotlinLogging
import org.lognet.springboot.grpc.GRpcService

private val log = KotlinLogging.logger {}

@GRpcService
class GrpcOrderCarService(
    private val orderService: OrderService,
    private val mapper: OrderMapper
) : CommonServiceGrpc.CommonServiceImplBase() {

    override fun createNewCarOrder(
        request: CommonDto.OrderDto,
        responseObserver: StreamObserver<CommonDto.OrderResultDto>
    ) {
        val orderDto = mapper.grpcOrderDtoToOrderDto(request)
        log.info { "Message '$orderDto' received" }
        try {
            orderService.createNewOrderByCarIdAndDescription(orderDto.carId, orderDto.orderDescription)
                .also {
                    log.info("Order '$it' saved")
                    responseObserver.onNext(mapper.toGrpcOrderResultDto(it.carId, EventStatus.SUCCESS))
                }
        } catch (ex: Exception) {
            log.info("Something went wrong... ${ex.message}")
            responseObserver.onNext(mapper.toGrpcOrderResultDto(orderDto.carId, EventStatus.FAILED))
        } finally {
            responseObserver.onCompleted()
        }
    }
}
